File: SocialChallenge/Managers/Parsers/Abstract/JSCParser.m
//
//  JSCParser.m
//  SocialChallenge
//
//  Created by Javier Laguna on 31/01/2016.
//
//

#import "JSCParser.h"

@interface JSCParser ()

/**
 Context for the parser to access CoreData.
 */
@property (nonatomic, strong, readwrite) NSManagedObjectContext *managedContext;

@end

@implementation JSCParser

#pragma mark - Parser

+ (instancetype)parserWithContext:(NSManagedObjectContext *)managedContext
{
    JSCParser *parser = [[self.class alloc] init];
    
    parser.managedContext = managedContext;
    
    return parser;
}

#pragma mark - Init

- (instancetype)init
{
    return [self initWithContext:nil];
}

- (instancetype)initWithContext:(NSManagedObjectContext *)managedContext
{
    self = [super init];
    
    if (self)
    {
        self.managedContext = managedContext;
    }
    
    return self;
}

@end

File: SocialChallenge/Managers/Parsers/Abstract/JSCParser.swift
//
//  JSCParser.swift
//  SocialChallenge
//
//  Created by Javier Laguna on 07/03/2016.
//
//

import Foundation
import CoreData

//MARK: ValueOrDefault

/**
Convenient method to check if a value is not nil and returns ir or the default
*/
@inline(__always) func JSCValueOrDefault(value: AnyObject?, defaultValue: AnyObject?) -> AnyObject?
{
    if value == nil ||
        value is NSNull {
            
            return defaultValue
    }
    
    return value
}

@objc(JSCParser)

/**
Code base for the parsers.
*/
class JSCParser: NSObject {
    
    /**
     Context for the parser to access CoreData.
     */
    var managedObjectContext = NSManagedObjectContext ()
    
    //MARK: Init
    
    /**
    Initialises the parser.
    
    - Parameter managedContext - Context for the parser to access CoreData.
    
    - Returns: FSNParser instance.
    */
    required init(managedObjectContext: NSManagedObjectContext?) {
        
        self.managedObjectContext = managedObjectContext!;
    }
    
    //MARK: Parser
    
    /**
    Convenient initialiser the parser.
    
    - Parameter managedContext - Context for the parser to access CoreData.
    
    - Returns: FSNParser instance.
    */
    class func parserWithContext(managedObjectContext: NSManagedObjectContext) -> Self {
        
        return self.init(managedObjectContext: managedObjectContext)
    }
}